src/librustc/infer/error_reporting.rs

```diff
@@ -1000,52 +1000,54 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         let parent_node = self.tcx.map.find(parent);
         let taken = lifetimes_in_scope(self.tcx, scope_id);
         let life_giver = LifeGiver::with_taken(&taken[..]);
-        let node_inner = match parent_node {
+        let (is_impl, node_inner) = match parent_node {
             Some(ref node) => match *node {
                 ast_map::NodeItem(ref item) => {
                     match item.node {
                         hir::ItemFn(ref fn_decl, unsafety, constness, _, ref gen, _) => {
-                            Some((fn_decl, gen, unsafety, constness, item.name, item.span))
-                        },
-                        _ => None
+                            (false, Some((fn_decl, gen, unsafety, constness, item.name, item.span)))
+                        }
+                        _ => (false, None),
                     }
                 }
                 ast_map::NodeImplItem(item) => {
                     match item.node {
                         hir::ImplItemKind::Method(ref sig, _) => {
-                            Some((&sig.decl,
+                            (true, Some((&sig.decl,
```

Comment:
I think you can just add an early return here and that might be enough. Also this should only be done if the `impl` is for a trait, not inherent.